pennylane/ops/op_math/condition.py

```diff
@@ -647,39 +665,134 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def _register_custom_staging_rule(cond_prim):
+
+    # see https://github.com/jax-ml/jax/blob/9e62994bce7c7fcbb2f6a50c9ef89526cd2c2be6/jax/_src/lax/lax.py#L3538
+    # and https://github.com/jax-ml/jax/blob/9e62994bce7c7fcbb2f6a50c9ef89526cd2c2be6/jax/_src/lax/lax.py#L208
+    # for reference to how jax is handling staging rules for dynamic shapes in v0.4.28
+
+    import jax  # pylint: disable=import-outside-toplevel
+    from jax._src.interpreters import partial_eval as pe  # pylint: disable=import-outside-toplevel
+
+    def _tracer_and_outvar(jaxpr_trace, outvar, env: dict):
+        """
+        Create a new tracer and returned var from the true branch outvar
+        returned vars are cached in env for use in future shapes
+        """
+        if not hasattr(outvar.aval, "shape"):
+            out_tracer = pe.DynamicJaxprTracer(jaxpr_trace, outvar.aval)
+            return out_tracer, jaxpr_trace.makevar(out_tracer)
+        new_shape = [s if isinstance(s, int) else env[s] for s in outvar.aval.shape]
+        new_aval = jax.core.DShapedArray(tuple(new_shape), outvar.aval.dtype)
+        out_tracer = pe.DynamicJaxprTracer(jaxpr_trace, new_aval)
+        new_var = jaxpr_trace.makevar(out_tracer)
+
+        if not isinstance(outvar, jax.core.Literal):
+            env[outvar] = new_var
+        return out_tracer, new_var
+
+    def custom_staging_rule(jaxpr_trace, *tracers, **params):
+        """
+        Add new jaxpr equation to the jaxpr_trace and return new tracers.
+        """
+        if not jax.config.jax_dynamic_shapes:
+            # fallback to normal behavior
+            return jaxpr_trace.default_process_primitive(cond_prim, tracers, params)
+        true_outvars = params["jaxpr_branches"][0].outvars
+
+        env = {}  # branch var to new equation var
+        out_tracers, returned_vars = tuple(
+            zip(*(_tracer_and_outvar(jaxpr_trace, var, env) for var in true_outvars))

```

Comment:
```suggestion
            zip(*(_tracer_and_outvar(jaxpr_trace, var, env) for var in true_outvars), strict=True)

```

Reply:
Adding locally to get the formatting right.

But not as critical here, since we know that `tracer_and_outvar` always returns two things for each input.

This is a fancy way to do
```
a = [(1,2), (3,4), (5,6)]
[x[0] for x in a], [x[1] for x in a]
```
We know those two variables will always be the same length, since they will always be the same length as `a`.

Reply:
You're right, I kind of blindly suggested it 😅